src/modules/personenkontext/specification/nur-gleiche-rolle.ts

```diff
@@ -0,0 +1,48 @@
+import { RolleRepo } from '../../rolle/repo/rolle.repo.js';
+import { DBiamPersonenkontextRepo } from '../persistence/dbiam-personenkontext.repo.js';
+import { RollenArt } from '../../rolle/domain/rolle.enums.js';
+import { Personenkontext } from '../domain/personenkontext.js';
+import { Rolle } from '../../rolle/domain/rolle.js';
+
+export class CheckRollenartSpecification {
+    public constructor(
+        private readonly personenkontextRepo: DBiamPersonenkontextRepo,
+        private readonly rolleRepo: RolleRepo,
+    ) {}
+
+    public async checkRollenart(sentPKs: Personenkontext<boolean>[]): Promise<boolean> {
+        // Fetch all personenkontexte for the person
+        let existingPKs: Personenkontext<true>[] = [];
+
+        if (sentPKs[0]?.personId) {
+            existingPKs = await this.personenkontextRepo.findByPerson(sentPKs[0]?.personId);
+        }
+
+        // Get roles for existing and sent PKs
+        const existingRollen: Rolle<true>[] = await this.getUniqueRollenFromPersonenkontexte(existingPKs);
+        const sentRollen: Rolle<true>[] = await this.getUniqueRollenFromPersonenkontexte(sentPKs);
+
+        // If there are no existing roles, any new roles are allowed
+        if (existingRollen.length === 0) {
+            return true;
+        }
+
+        // Get the RollenArt of existing PKs (assuming all existing PKs have the same RollenArt)
+        const existingRollenArt: RollenArt | undefined = existingRollen[0]?.rollenart;
+
+        // Check if all sent roles match the existing RollenArt
+        const allSentRollenMatchExisting: boolean = sentRollen.every(
+            (rolle: Rolle<true>) => rolle.rollenart === existingRollenArt,
+        );
+
+        return allSentRollenMatchExisting;
+    }
+
```

Comment:
Für den Fall, dass personenkontextRepo.findByPerson ein leeres Array liefert oder "sentPKs[0]?.personId == null" ist, bleibt existingPKs ein leeres Array. Da damit getUniqueRollenFromPersonenkontexte() auch ein leeres Array zurückliefert ist die Bedingung "existingRollen.length === 0" immer erfüllt und die Spezifikation gibt true zurück auch dann wenn sentPKs zwei Personenkontexte mit unterschiedlicher Rollenart enthält...

---> Selbst wenn es aktuell keine existierenden PKs gibt müssen trotzdem die neuen PK's in sich auf Gültigkeit geprüft werden, oder?

Reply:
Das sollte normalerweise nie vorkommen, wenn man über die UI Kontexte hinzufügt, weil man nur 1 Kontext pro Aktion hinzufügen kann aber du hast Recht, der API muss trotzdem gesichert werden.